### PR TITLE
Introduce counters map

### PR DESCRIPTION
**Pull request by vlrpl**
For the time being, its purpose is to carry event errors.
The series includes:

- UAF fix (#151)
- detach capabilities to all the bpfs (this would stop the counters manipulation)
- add error_report flag in rodata for fast access (atm tied to --debug-ebpf, but could be replaced, or removed altogether, depending on the preference)
- actual event error report (#128)
- optimization that reduces the complexity (could be dropped, squashed, kept separate)

the series is ready to be reviewed, but of course requires some decisions mostly about the flag (whether we want it or not, and if so, a new one or an old is fine) and the optimization.

Below some simulated loss output:

```
09:00:20 [WARN] lost: 12 at ovs_execute_actions                                                                                                                                                                                             
09:00:20 [WARN] lost: 6 at ovs_dp_upcall                                                                                                                                                                                                    
09:00:20 [WARN] lost: 6 at /usr/local/sbin/ovs-vswitchd:dpif_netlink_operate__:op_flow_execute                                                                                                                                              
09:00:20 [WARN] lost: 6 at openvswitch:ovs_dp_upcall                                                                                                                                                                                        
09:00:20 [WARN] lost: 35 at net:netif_receive_skb                                                                                                                                                                                           
09:00:20 [WARN] lost: 6 at /usr/local/sbin/ovs-vswitchd:dpif_netlink_operate__:op_flow_execute                                                                                                                                              
09:00:20 [WARN] lost: 1401 at skb_free_head                                                                                                                                                                                                 
09:00:20 [WARN] lost: 6 at skb:kfree_skb                                                                                                                                                                                                    
09:00:20 [WARN] lost: 6 at queue_userspace_packet                                                                                                                                                                                           
09:00:20 [WARN] lost: 12 at openvswitch:ovs_do_execute_action                                                                                                                                                                               
09:00:20 [WARN] lost: 183 at pskb_expand_head                                                                                                                                                                                               
09:00:20 [WARN] lost: 19 at net:netif_rx                                                                                                                                                                                                    
09:00:20 [WARN] total events lost: 1686                                                                                                                                                                                                     
```

### CODE REVIEW

File: src/core/probe/kernel/kprobe.rs [hunk 1]
```diff
@@ -17,8 +17,8 @@ use kprobe_bpf::KprobeSkelBuilder;
 
 #[derive(Default)]
 pub(crate) struct KprobeBuilder {
-    obj: Option<libbpf_rs::Object>,
     links: Vec<libbpf_rs::Link>,
+    obj: Option<libbpf_rs::Object>,
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/kprobe.rs:21]: ack for the kretprobe builder.
For the PR, yes, at least as it is.
Didn't dig much into that as the pr was more narrowed toward its goals, the idea was to expand later as we don't have any observable issue with those, but I guess that being a bit conservative is ok. In general, I think the underlying library was not conceived with the early drop in mind, so keeping the reference makes sense. I don't consider it defensive in this case.

File: src/core/probe/manager.rs [hunk 2]
```diff
@@ -241,6 +259,20 @@ impl ProbeManager {
         Ok(())
     }
 
+    /// Detach all probes.
+    pub(crate) fn detach(&mut self) -> Result<()> {
+        self.builders
+            .iter_mut()
+            .try_for_each(|builder| -> Result<()> { builder.detach() })?;
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/manager.rs:266]: thanks. This slipped out among the various local revisions.

File: src/core/probe/bpf/include/common_defs.h [hunk 3]
```diff
@@ -1,7 +1,56 @@
...
+/* Contains the counters of the error path.  This is then processed
+ * and reported from user-space. */
+struct retis_counters {
+	u64 events;
+};
+
+/* Probe configuration; the key is the target symbol address */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, PROBE_MAX * 2);
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/bpf/include/common_defs.h:34]: PROBE_MAX was thought for kernel probes.
I initially doubled it because it still was a kernel entity. We should probably give it a fixed value without doubling it and use a different value for the kernel max number of probes. 
This map should take into account KERNEL_PROBES_MAX + USDT_PROBES_MAX although I expect USDTs to remain fairly contained in terms of numbers
WDYT?

File: src/core/probe/manager.rs [hunk 4]
```diff
@@ -304,6 +348,62 @@ impl ProbeManager {
...
+                counters
+                    .copy_from_bytes(&counters_val)
+                    .or_else(|_| bail!("Cannot retrieve the counters map value"))?;
+                if counters.events == 0 {
+                    continue;
+                }
+
+                /* kernel symbols */
+                if counters_key.pid == 0 {
+                    let ksym = Symbol::from_addr(counters_key.sym_addr)?;
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/manager.rs:377]: > * I don't think the initial version could distinguish between kprobes and kretprobes, because both Symbol would return the same symbol address and the first one found would match. This is actually better because not reporting the probe type is not misleading.

yes, that's true, I probably needed some coffee when I went through it. I also don't love the idea to carry that information in the map (e.g. a selector), which was one of the ideas involved.

> 
>     * Having said that, we should fix this one day. Please add an issue to track it.

sure

> 
>     * Suggestion: building `Symbol` is not completely free either (but not bad IMHO) so if we want to further improve perfs we could traverse the probes list once and build a map indexed by the symbol addresses.

makes sense, and it was an option while writing this. It should not take much time, but I don't have plenty. If ok, we could keep this as best effort and possibly add an issue aimed to improve this part if it doesn't make it into this round.
WDYT?

File: src/core/probe/bpf/include/common_defs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/bpf/include/common_defs.h:34]: > > PROBE_MAX was thought for kernel probes.
> 
> No this is actually the global limit, see `check_probe_max` in the probe manager. The kernel config map do use it as it's the upper limit, but it's not kernel specific.

ah, right. My memory failed. I was sure it was defined under the kernel probe part (which seemed a bit weird, btw), but that simply was not true. So, '*2' doesn't make any sense and just need to be dropped

File: src/core/probe/kernel/bpf/include/common.h [hunk 5]
```diff
@@ -196,12 +196,19 @@ static __always_inline void filter(struct retis_context *ctx)
 	ctx->filters_ret |= (!!fctx.ret) << RETIS_F_PACKET_PASS_SH;
 }
 
+static __always_inline void event_err(u64 addr, struct retis_raw_event *e)
+{
+	err_report(addr, 0);
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:201]: didn't do it as in that case discard_event() would need pid and symbol.
What about dropping the helper and moving err_report() in the discard_event() call site?
I guess we'll always stay (or at least we should try :) with a centralized err path for event discarding, so this should be enough.
WDYT?

File: src/core/probe/bpf/include/common_defs.h [hunk 6]
```diff
@@ -1,7 +1,56 @@
...
+#include <events.h>
+
+/* Keep in sync with its Rust counterpart in crate::core::probe */
+#define PROBE_MAX	1024
+
 #define COMMON_SECTION_CORE	0
 #define COMMON_SECTION_TASK	1
 
+/* Error reporting is enabled. */
+const volatile u8 report_error = 0;
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/bpf/include/common_defs.h:15]: this was part of the decision I was referring to in the PR descr.
If we agree to always report, one patch should be dropped as well.
I'm fine with it, though.

File: src/core/probe/common.rs [hunk 7]
```diff
@@ -32,3 +55,20 @@ pub(crate) fn get_ebpf_debug() -> bool {
...
+        sz: std::mem::size_of::<libbpf_sys::bpf_map_create_opts>() as libbpf_sys::size_t,
+        ..Default::default()
+    };
+
+    Ok(libbpf_rs::Map::create(
+        libbpf_rs::MapType::Hash,
+        Some("counters_map"),
+        std::mem::size_of::<CountersKey>() as u32,
+        std::mem::size_of::<Counters>() as u32,
+        (PROBE_MAX * 2) as u32,
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/common.rs:71]: grr  :)

File: src/core/probe/kernel/bpf/include/common.h [hunk 8]
```diff
@@ -244,7 +253,7 @@ static __always_inline int chain(struct retis_context *ctx)
 	if (!ti)
 		goto discard_event;
 
-	ti->pid = bpf_get_current_pid_tgid();
+	ti->pid = pid;
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:256]: it saves a helper call (maybe it gets optimized). Would you prefer to call it anyways?

[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:256]: sorry, you're right, doesn't save anything, it does for usdt only

File: src/core/probe/kernel/bpf/include/common.h [hunk 9]
```diff
@@ -286,11 +295,13 @@ static __always_inline int chain(struct retis_context *ctx)
 	CALL_HOOK(8)
 	CALL_HOOK(9)
 
-	if (get_event_size(event) > pass_threshold)
+	if (get_event_size(event) > pass_threshold) {
 		send_event(event);
-	else
+	} else {
 discard_event:
+		err_report(ctx->ksym, 0);
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:302]: > This doesn't work because there are valid cases to jump here:
> 
>     * Hook filtered the event.

while giving a careful look at it, I noticed something that I missed in a previous review here.
Unfortunately, we cannot rely on -ENOMSG (I guess in skb based modules) because attaching the program to a tracepoint with hooks returning non-zero value results in:

```
At program exit the register R0 has value (0xffffffd6; 0x0) should have been in (0x0; 0x0)
processed 901 insns (limit 1000000) max_states_per_insn 0 total_states 47 peak_states 47 mark_read 41
-- END PROG LOAD LOG --
libbpf: prog 'hook': failed to load: -22
libbpf: failed to load object 'hook'
Error: System error, errno: 22

```

Filtering worked around it. I think we should file an issue for this.

[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:302]: given that we know sections will succeed and are not counted for hooks, and given that we can't rely on hooks retval (because of the previous comment), I think the best thing to do is to narrow the scope of the counters to its actual and initial purpose (ringbuf exhaustion) and report to user-space if get_event() failed. Other types of error will be added (event sections errors, if they can happen). Furthermore some might be better suited for some kind of dedicated trace pipe rather than counters and may benefit more from that kind of error reporting facility.

[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:302]: I would turn dropped_events into ringbuf_errors and change the warn!().
`{} ring buffer error(s) in {}`

File: src/core/probe/user/user.rs [hunk 10]
```diff
@@ -45,6 +49,17 @@ impl UsdtProbe {
     pub(crate) fn name(&self) -> String {
         format!("usdt:{}:{}", self.provider, self.name)
     }
+
+    /// Generate the probe BPF configuration from a list of options.
+    pub(crate) fn gen_counters(&self, _options: &[ProbeOption]) -> Result<(CountersKey, Counters)> {
```
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/user/user.rs:54]: it was meaningful initially, but then later lost its meaning and just remained for consistency with gen_config().
We'll probably never end up using those options there (and we can add them if we will).
I'll remove the parameter.

File: src/core/probe/kernel/bpf/include/common.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (vlrpl) [reply src/core/probe/kernel/bpf/include/common.h:302]: ack, works for me
